cloudmarker/alerts/emailalert.py

```diff
@@ -12,98 +12,86 @@
+        """Save event record in a buffer.
 
         Arguments:
-            record (dict): Data to send via email.
+            record (dict): An event record.
         """
         for _, value in record.items():
             self._stringbuffer.append(repr(value))
 
     def done(self):
-        """Perform final cleanup tasks.
-
-        This method is called after all records have been written. In
-        this example implementation, we properly terminate the JSON
-        array in the email body.
-
-        The connection is made based whether the SMTP or SMTP_SSL is
-        required which is determined by ``use_ssl`` param. The SMTP
-        login process is conducted if ``username`` param is provided.
-
-        """
+        """Send the buffered events as an email alert."""
         message = MIMEMultipart()
         message['Date'] = email.utils.formatdate(localtime=True)
         message['Subject'] = self._subject
-        message['From'] = self._sender
-        message['To'] = email.utils.COMMASPACE.join(self._to)
+        message['From'] = self._from_addr
+        message['To'] = email.utils.COMMASPACE.join(self._to_addrs)
 
-        # Incase if the buffer is empty then the default message will be sent
-        if self._stringbuffer:
-            self._body = ''.join(self._stringbuffer)
-        message.attach(MIMEText(self._body))
+        body = ''.join(self._stringbuffer)
+        message.attach(MIMEText(body))
 
         smtp = self._prepare_smtp_session()
         try:
-            smtp.sendmail(self._sender, self._to, message.as_string())
+            smtp.sendmail(self._from_addr,
+                          self._to_addrs,
+                          message.as_string())
         except smtplib.SMTPException as e:
             _log.error('Failed to send email: %s', e)
         finally:
             smtp.quit()
 
     def _prepare_smtp_session(self):
-        """Return SMTP connection object.
-
-        Create a SMTP connection based on whether the SSL param.
-        If SSL connection is required (for eg: in case of gmail) then
-        method will return a SMTP_SSL connection. In other cases if the
-        authentication is not required or SMTP_SSL connection is not
-        required then a plain SMTP connection object is returned.
-        For gmail follow the steps given below even to work for 2 factor
-        authentication
-        1. Log-in into Gmail with your account
-        2. Navigate to https://security.google.com/settings/security/
-        apppasswords
-        3. In 'select app' choose 'custom', give it an arbitrary name and
-        press generate
-        4. It will give you 16 chars token.
-        5. Use that token as password for login
-        6. Host is smtp.gmail.com and port is 465
+        """Start an SMTP session.
+
+        Returns:
+            smtplib.SMTP or smtplib.SMTP_SSL: An SMTP connection
+                with its session ready to send messages.
 
         """
         if self._use_ssl:
```

Comment:
In lines [99](https://github.com/cloudmarker/cloudmarker/pull/98/commits/ba56330d1b85174f7258121ce04b481267580cec#diff-28d4faa43f2d620f5a2c54123370ee47R99) and [102](https://github.com/cloudmarker/cloudmarker/pull/98/commits/ba56330d1b85174f7258121ce04b481267580cec#diff-28d4faa43f2d620f5a2c54123370ee47R102), enabling the debug level by default leads to a lot of email body data in the logs. Should this flag be moved to the config as an internal debug parameter?

Reply:
At line [98](https://github.com/cloudmarker/cloudmarker/pull/98/commits/ba56330d1b85174f7258121ce04b481267580cec#diff-28d4faa43f2d620f5a2c54123370ee47R98), with the following config, I ran into an `ssl.SSLError`.

Config:
```json
alerts:
  emailalert:
    params:
      use_ssl: True
      host: smtp.gmail.com
      port: 587
      from_addr: CloudMarker Alert <no-reply@gmail.com>
      to_addrs:
        - <recipient@example.com>
      subject: Anomaly notification
      username: <alice>
      password: <alice@123>
```

Exception stack trace:
```
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/git/cloudmarker/cloudmarker/workers.py", line 68, in store_worker
    store_plugin.done()
  File "/Users/git/cloudmarker/cloudmarker/alerts/emailalert.py", line 79, in done
    smtp = self._prepare_smtp_session()
  File "/Users/git/cloudmarker/cloudmarker/alerts/emailalert.py", line 98, in _prepare_smtp_session
    smtp = smtplib.SMTP_SSL(host=self._host, port=self._port)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/smtplib.py", line 1031, in __init__
    source_address)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/smtplib.py", line 251, in __init__
    (code, msg) = self.connect(host, port)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/smtplib.py", line 336, in connect
    self.sock = self._get_socket(host, port, self.timeout)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/smtplib.py", line 1039, in _get_socket
    server_hostname=self._host)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 412, in wrap_socket
    session=session
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 853, in _create
    self.do_handshake()
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 1117, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:1056)
```

If the `use_ssl` key is set to `False`, the mail is successfully sent. This could probably be handled in another PR. There is a solution mentioned [here](https://stackoverflow.com/a/49443966).

Reply:
> In lines [99](https://github.com/cloudmarker/cloudmarker/pull/98/commits/ba56330d1b85174f7258121ce04b481267580cec#diff-28d4faa43f2d620f5a2c54123370ee47R99) and [102](https://github.com/cloudmarker/cloudmarker/pull/98/commits/ba56330d1b85174f7258121ce04b481267580cec#diff-28d4faa43f2d620f5a2c54123370ee47R102), enabling the debug level by default leads to a lot of email body data in the logs. Should this flag be moved to the config as an internal debug parameter?

I agree. This should be config driven. I will do this as part of #53.

Reply:
> At line [98](https://github.com/cloudmarker/cloudmarker/pull/98/commits/ba56330d1b85174f7258121ce04b481267580cec#diff-28d4faa43f2d620f5a2c54123370ee47R98), with the following config, I ran into an `ssl.SSLError`.
> 
> Config:
> 
> ```json
> alerts:
>   emailalert:
>     params:
>       use_ssl: True
>       host: smtp.gmail.com
>       port: 587

For `use_ssl` set to `True`, set `port` to either `465` or `0` (to choose default port which is `465`).

Also, see the docstring in [line 26](https://github.com/cloudmarker/cloudmarker/pull/98/files#diff-28d4faa43f2d620f5a2c54123370ee47R26) for more details on this.

Reply:
The docstring does not say that `use_ssl` can only be used for port 465. I'm not sure if someone running into this error would be able to figure out that both the `port` and `use_ssl` work together. This is not a breaking change so I'm fine if we don't do error handling here.

Reply:
I tried to address this by adding more details in the `util.send_email()` docstring in PR #108. See the following two URLs on the Read the Docs:

- [cloudmarker.alerts.emailalert.EmailAlert()](https://cloudmarker.readthedocs.io/en/latest/api/cloudmarker.alerts.html#module-cloudmarker.alerts.emailalert)
- [cloudmarker.util.send_email()](https://cloudmarker.readthedocs.io/en/latest/api/cloudmarker.html#cloudmarker.util.send_email)

Please let me know if this helps of if we need to rework on these docstrings.